packages/grpc-js/src/server-call.ts

```diff
@@ -146,3 +246,209 @@ export type Handler<RequestType, ResponseType> = {
 };
 
 export type HandlerType = 'bidi'|'clientStream'|'serverStream'|'unary';
+
+
+// Internal class that wraps the HTTP2 request.
+export class Http2ServerCallStream<RequestType, ResponseType> extends
+    EventEmitter {
+  cancelled = false;
+  deadline: NodeJS.Timer|null = null;
+  status: PartialServiceError = {code: Status.OK, details: 'OK'};
+
+  constructor(
+      private stream: http2.ServerHttp2Stream,
+      private handler: Handler<RequestType, ResponseType>|null) {
+    super();
+
+    this.stream.once('error', (err: Error) => {
+      this.sendError(err as ServiceError, Status.INTERNAL);
+    });
+
+    this.stream.once('close', () => {
+      if (this.stream.rstCode === http2.constants.NGHTTP2_CANCEL) {
+        this.cancelled = true;
+        this.emit('cancelled', 'cancelled');
+      }
+    });
+  }
+
+  sendMetadata(customMetadata?: Metadata) {
+    if (this.stream.headersSent === true) {
+      return;
+    }
+
+    const custom = customMetadata ? customMetadata.toHttp2Headers() : null;
+    // TODO(cjihrig): Include compression headers.
+    const headers = Object.assign(defaultResponseHeaders, custom);
+
+    this.stream.once('wantTrailers', () => {
+      let trailersToSend = {
+        [kGrpcStatusHeader]: this.status.code,
+        [kGrpcMessageHeader]: encodeURI(this.status.details as string)
+      };
+      const metadata = this.status.metadata;
+
+      if (metadata) {
+        trailersToSend =
+            Object.assign(trailersToSend, metadata.toHttp2Headers());
+      }
+
+      this.stream.sendTrailers(trailersToSend);
+    });
+    this.stream.respond(headers, defaultResponseOptions);
```

Comment:
Why is this line after the `'wantTrailers'` event listener? That makes it easy to miss that this line is in the `sendMetadata` scope.